sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -1018,9 +1018,20 @@ case class TimeAdd(start: Expression, interval: Expression, timeZoneId: Option[S
 }
 
 /**
- * Given a timestamp like '2017-07-14 02:40:00.0', interprets it as a time in UTC, and renders
- * that time as a timestamp in the given time zone. For example, 'GMT+1' would yield
- * '2017-07-14 03:40:00.0'.
+ * This is a common function for databases supporting TIMESTAMP WITHOUT TIMEZONE. This function
+ * takes a timestamp which is timezone-agnostic, and interprets it as a timestamp in UTC, and
+ * renders that timestamp as a timestamp in the given time zone.
+ *
+ * However, timestamp in Spark represents number of microseconds from the Unix epoch, which is not
+ * timezone-agnostic. So in Spark this function just shift the timestamp value from UTC timezone to
+ * the given timezone.
+ *
+ * This function may return confusing result if the input is a string with timezone, e.g.
+ * '2018-03-13T06:18:23+00:00'. The reason is that, Spark firstly cast the string to timestamp
+ * according to the timezone in the string, and finally display the result by converting the
+ * timestamp to string according to the session local timezone.
+ *
+ * We may remove this function in Spark 3.0.
```

Comment:
Hm, how about explicitly deprecating them in Scala/Python/R APIs?

Reply:
It's too late for 2.4, I'm thinking about deprecating them in 2.5. @rxin what do you think?

Reply:
should also update the sql doc?

Reply:
Hmmm .. I think we should get rid of this sentence here and just formally deprecate it in 2.5.0 ..  don't feel strongly.

Reply:
should also update other language APIs as well.

Reply:
This may be too long to put into the SQL doc, I'll update doc in other languages.

Reply:
I thought this is a simple documentation change for SQL function doc for 2.4.0 and https://issues.apache.org/jira/browse/SPARK-25496 is aiming the official global deprecation at 2.5.0.

Reply:
If we need to do `deprecation` for all, let's update the PR title clearly; maybe additional `[SPARK-25496]` and `Deprecation` wording at the title.

Reply:
Deprecation is kind of a warning before removal .. warning for deprecation sounds like a warning for a warning ..

Reply:
What I meant is `Deprecation` is a more formal activity requiring complete coverage. And, it's open as SPARK-25496.

Reply:
The decision is not made yet, anyway I won't deprecate it in this PR, we definitely need a new PR